sympy/solvers/solveset.py

```diff
@@ -978,6 +985,342 @@ def _solveset(f, symbol, domain, _check=False):
+    {-log(375)/(-log(5) + 2*log(3))}
+
+    How to use `\_transolve`
+    ========================
+
+    `\_transolve` should be used as an internal function. There should be
+    an internal call to `\_transolve`, passing all the three above
+    parameters as arguments. These three arguments should be sympy
+    instances.
+
+    How `\_transolve` works
+    =======================
+
+    `\_transolve` uses two types of helper functions to solve equation
+    of a particular class:
+
+    Identifying helpers: These helpers are used to determine whether a
+    given equation belongs to a certain class of equation or not.
+    These functions return either True or False.
+
+    Solving helpers: These are the helpers that solve the equation. They
+    are invoked once the equation is identified by the `identifying
+    helpers`. These function return either the complete solution or
+    modifies the equation for `solveset` to handle.
+
+    * Philosophy behind the module
+
+    `\_transolve` comes into action when `solveset` is unable to solve
+    the equation as a last resort to get the solutions. The idea is to
+    get the equation in a case (Add, Pow etc.) that satisfies its
+    general form, for example exponential equations take the form as
+    `a*f(x) + b*g(x)`, where `f(x)` and `g(x)` are power terms,
+    therefore they are included in `Add` case.
+
+    Once done different `identifying helper` conditions are included
+    to check which class the equation belongs. `Solving helper` is
+    invoked once the class of the equation is identified. If the
+    equation is solved the result is returned otherwise a
+    `ConditionSet` is returned.
+
+    How `\_transolve` is better than `\_tsolve`
+    ===========================================
+
+    1) Better output
+
+    `\_transolve` provides a better output to some equations than
+    `\_tsolve`. Though result from both the functions are correct
+    it's just that the one from `\_transolve` is easy to understand and
+    appropriate. This is because `\_transolve` simplifies the result
+    wherever necessary.
+
+    Consider a simple exponential equation
+
+    >>> f = 3**(2*x) - 2**(x + 3)
+    >>> pprint(transolve(f, x, S.Reals), use_unicode=False)
+        -3*log(2)
+    {------------------}
+     -2*log(3) + log(2)
+    >>> pprint(tsolve(f, x), use_unicode=False)
+         /   3     \
+         | --------|
+         | log(2/9)|
+    [-log\2         /]
+
+    `\_transolve's` output seems more simple, readable and appropriate
+    as compared to that of `\_tsolve`.
+
+    2) Extensible
+
+    The API of `\_transolve` is designed such that it is easily extensible,
+    i.e, it is easy to add a new class of equation solver without having
+    to mess with the API or solving code of other helpers. The idea
+    behind `\_transolve` is making seperate helpers for different equation
+    solver. So to make new class of equation solve, you must add a
```

Comment:
So to... -> So to solve a given class of equation, one helper identifies the equation and another solves the equation or modifies it into a form that can be solved.

Actually, I don't think the helper is solving the equation, it is just recasting the equation into a tractable form, isn't it?

Reply:
> Actually, I don't think the helper is solving the equation, it is just recasting the equation into a tractable form, isn't it?

The thought is helpers can do either thing, convert into a tractable form or solve completely. Like
`expo_solver` reduces the equation whereas `log solver` would return the solution (atleast as of now)

Reply:
Where is log solver? 

Would it make sense to name the helpers like those that are already present in solveset: `_solve_abs`, `_solve_trig`, etc...?

Reply:
Is `_solve_transcendental` really extensible? What other sorts of equations besides 2 (or more) exponentials will it solve? Should it maybe just be `_solve_exp`?

Reply:
> Would it make sense to name the helpers like those that are already present in solveset: _solve_abs, _solve_trig, etc...?

Yes this can be done. Thanks!

Reply:
> Is _solve_transcendental really extensible? What other sorts of equations besides 2 (or more) exponentials will it solve? Should it maybe just be _solve_exp?

Other sorts of equations will include, logarithmic equations, lambert type equations and maybe we can incorporate solving by `solve_decomposition`.

Reply:
> The thought is helpers can do either thing, convert into a tractable form or solve completely

I commented on the log solver PR. IMO you *do not* want to go down this road since this will lead to decentralization. Keep the checking and domain clipping all in one place. Consider making transolve do the following:
1. step through an identify/transform sequence
2. if no identifications were a hit return the original equation or a ConditionSet?
3. if an identification is made return the changed form
4. if the call to transolve gets the same equation back it knows it wasn't a transcendental; if it is changed, `_solveset` passes the equation back to `_solveset` again.

...something like that.

Reply:
Yeah, this sounds better. I will keep this in mind while implementing other solvers.